### PR TITLE
Content: Unit test for pet-shop contract in javascript using chai 

### PR DESCRIPTION
**Pull request by AemieJ**
This PR has been made in regard to the suggestion added in  ```Suggestion : Test to be written in javascript as well #594```. The unit test has been written in javascript using chai module and made use of the asynchronous property to test the functionality of each function in ```pet-shop.md```.

### CODE REVIEW

**Comment by AemieJ** [comment]
@adrianmcli could you review this and let me know if it's good as I think both the tests are required.

**Comment by adrianmcli** [comment]
I don't work at Truffle anymore. Good luck!

**Comment by eggplantzzz** [comment]
Hey @AemieJ! Thanks for putting in the work for this PR. I like the idea for adding this, let me talk to the team to see what they think and try and ascertain the best way to integrate this if we move forward. Thanks again!

**Comment by eggplantzzz** [comment]
So @AemieJ, I talked with a bunch of the team and they were really into including a section on JS testing! I have an idea for how to include this in an organic way and there are some revisions that we would want to make. Would you be willing to go through a couple of rounds of review and revisions to get this merged? If not then I totally understand and can take it from here. Let me know if you want to work a bit more on this!

**Comment by AemieJ** [comment]
@eggplantzzz Yes, I wouldn't mind that!

**Comment by eggplantzzz** [comment]
Great! So my concept was to have the two sections collapsed by default on the tutorial and you can expand the one you are interested in. Here is a screenshot of what I imagine.

Screen Shot 2020-07-23 at 1 28 42 PM

You can do this by using an html `<details>` and `<summary>` tag. I think this would be a really nice way to do this. I'll go ahead and make some more comments a bit later.

File: src/tutorials/pet-shop.md [hunk 1]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
 
 Note the **memory** attribute on `adopters`. The memory attribute tells Solidity to temporarily store the value in memory, rather than saving it to the contract's storage. Since `adopters` is an array, and we know from the first adoption test that we adopted pet `expectedPetId`, we compare the testing contracts address with location `expectedPetId` in the array.
 
+## Testing the smart contract using Javascript 
+
+Truffle is very flexible when it comes to smart contract testing, in that tests can be written either in JavaScript or Solidity. In this tutorial, we'll be writing our tests in Javascript using chai and mocha library.
+
+1. Create a new file named `TestAdoption.test.js` in the `test/` directory.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:382]: So I think I would rather this test file be named `testAdoption.test.js` as I think it is more common to use Pascal case for files that export classes.

File: src/tutorials/pet-shop.md [hunk 2]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  To test the ```adopt()``` function, recall that upon success it returns the given ```petId```. We can ensure an ID was returned and that it's correct by comparing the return value of ```adopt()``` to the ID we passed in.
+
+  1. Add the following function within the ```TestAdoption.test.js``` test file, after the declaration of ```describe```. 
+
+  ```
+  describe('Test of functions', async() => {
+      const address = await Adoption.address; 
+      const expectedAdopter = accounts[0];
+
+      it("testUserCanAdoptPet", async() => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:428]: All the strings provided as the first argument for the `it` function should be written in plain English prose as if it were being started with "it". So this one should probably be something like `it("allows users to adopt a pet", async ()...`.

File: src/tutorials/pet-shop.md [hunk 3]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
 
+## Testing the smart contract using Javascript 
+
+Truffle is very flexible when it comes to smart contract testing, in that tests can be written either in JavaScript or Solidity. In this tutorial, we'll be writing our tests in Javascript using chai and mocha library.
+
+1. Create a new file named `TestAdoption.test.js` in the `test/` directory.
+1. Add the following content to the `TestAdoption.test.js` file:
+
+  ```
+  const Adoption = require("../contracts/Adoption.sol");
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:386]: This should use the globally available `artifacts` object. So you would do
```
const Adoption = artifacts.require("Adoption");
```

As you see you use the `require` method and give it the name of the contract.

File: src/tutorials/pet-shop.md (hunk 1 above)
Comment (eggplantzzz) [reply src/tutorials/pet-shop.md:382]: I wonder also if we should remove the `.test` portion of the filename. None of our other boxes use this convention for the JS tests. Do you feel strongly either way about it?

File: src/tutorials/pet-shop.md [hunk 4]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
 
 Note the **memory** attribute on `adopters`. The memory attribute tells Solidity to temporarily store the value in memory, rather than saving it to the contract's storage. Since `adopters` is an array, and we know from the first adoption test that we adopted pet `expectedPetId`, we compare the testing contracts address with location `expectedPetId` in the array.
 
+## Testing the smart contract using Javascript 
+
+Truffle is very flexible when it comes to smart contract testing, in that tests can be written either in JavaScript or Solidity. In this tutorial, we'll be writing our tests in Javascript using chai and mocha library.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:380]: The second "Javascript" should have the second "s" capitalized as it is in the first sentence.

File: src/tutorials/pet-shop.md [hunk 5]
```diff
@@ -277,7 +277,7 @@ Now we are ready to create our own migration script.
 You've now written your first smart contract and deployed it to a locally running blockchain. It's time to interact with our smart contract now to make sure it does what we want.
 
 
-## Testing the smart contract
+## Testing the smart contract using Solidity
 
 Truffle is very flexible when it comes to smart contract testing, in that tests can be written either in JavaScript or Solidity. In this tutorial, we'll be writing our tests in Solidity.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:282]: Perhaps "in this tutorial" should now be changed to "in this section, we'll write our tests in Solidity."

File: src/tutorials/pet-shop.md (hunk 4 above)
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:380]: Same as the comment above about "in this tutorial" being changed to "in this section".

File: src/tutorials/pet-shop.md [hunk 6]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  require('chai')
+    .use(require('chai-as-promised'))
+    .should()
+  
+  contract("TestAdoption", (accounts) => {
+    let adoption;
+    let expectedPetId = 8;
+    
+    before( async()=>{
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:396]: There are some odd spacings here. So this should be 
`before(async () => {`

File: src/tutorials/pet-shop.md [hunk 7]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+        adoption = await Adoption.deployed();
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:397]: It is conventional in Truffle JS code to use 2 space indentation and not 4.

File: src/tutorials/pet-shop.md [hunk 8]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+1. Add the following content to the `TestAdoption.test.js` file:
+
+  ```
+  const Adoption = require("../contracts/Adoption.sol");
+
+  require('chai')
+    .use(require('chai-as-promised'))
+    .should()
+  
+  contract("TestAdoption", (accounts) => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:392]: I think this should just read `contract("Adoption",...`

When you are running the test, that way it will just print out the contract name preceding the test results.

File: src/tutorials/pet-shop.md [hunk 9]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  ### Testing the adopt() function 
+
+  To test the ```adopt()``` function, recall that upon success it returns the given ```petId```. We can ensure an ID was returned and that it's correct by comparing the return value of ```adopt()``` to the ID we passed in.
+
+  1. Add the following function within the ```TestAdoption.test.js``` test file, after the declaration of ```describe```. 
+
+  ```
+  describe('Test of functions', async() => {
+      const address = await Adoption.address;
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:425]: Same comment about 2 space indentation.

File: src/tutorials/pet-shop.md [hunk 10]
```diff
@@ -375,6 +375,99 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  1. Add the following function within the ```TestAdoption.test.js``` test file, after the declaration of ```describe```. 
+
+  ```
+  describe('Test of functions', async() => {
+      const address = await Adoption.address; 
+      const expectedAdopter = accounts[0];
+
+      it("testUserCanAdoptPet", async() => {
+        const returnedId = await adoption.adopt(expectedPetId, {from: accounts[0]});
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:429]: I think it is convention in Truffle code to add spaces before braces and the contents like `{ from: accounts[0] })` and not `{from: accounts[0]})`

**Comment by eggplantzzz** [comment]
So I reworked the test itself a little bit and tried to make it clean and concise. What do you think about this for this as the finished test?
```
const Adoption = artifacts.require("Adoption");

contract("Adoption", (accounts) => {
  let adoption;
  let expectedAdopter;

  before(async () => {
    adoption = await Adoption.deployed();
  });

  describe("adoptioning and retrieving owners' accounts", async () => {
    before("adopt a pet using accounts[0]", async () => {
      await adoption.adopt(8, { from: accounts[0] });
      expectedAdopter = accounts[0];
    });

    it("can fetch the address of an owner by pet id", async () => {
      const adopter = await adoption.adopters(8);
      assert.equal(adopter, expectedAdopter, "The owner of the adopted pet should be the first account.");
    });

    it("can fetch the collection of all pet owners addresses", async () => {
      const adopters = await adoption.getAdopters();
      assert.equal(adopters[8], expectedAdopter, "The owner of the adopted pet should be in the collection.");
    });
  });
});
```

File: src/tutorials/pet-shop.md (hunk 1 above)
Comment (AemieJ) [reply src/tutorials/pet-shop.md:382]: It's not necessary to put `.test` within the fileName however, when one reads their code, it is easier to differentiate between the test files from the non-test files.

**Comment by AemieJ** [comment]
@eggplantzzz it has the same logic so yes it's perfect except I think for javascript you need to include chai as well for performing the tests.

File: src/tutorials/pet-shop.md (hunk 1 above)
Comment (eggplantzzz) [reply src/tutorials/pet-shop.md:382]: Sure, I'm more concerned about the fact that it breaks from convention that we use elsewhere. But if you want to keep it in then that is ok with me!

**Comment by eggplantzzz** [comment]
Actually you don't need to import `chai` into your tests. Truffle automatically includes it for you and makes `assert` available in the test environment!

**Comment by AemieJ** [comment]
> Great! So my concept was to have the two sections collapsed by default on the tutorial and you can expand the one you are interested in. Here is a screenshot of what I imagine.
> 
> Screen Shot 2020-07-23 at 1 28 42 PM
> 
> You can do this by using an html `<details>` and `<summary>` tag. I think this would be a really nice way to do this. I'll go ahead and make some more comments a bit later.

For this case, if I include `<details>` and `<summary>` tag to put a collapsible section in the markdown, I have to include each and every specific HTML element in the markdown as well. If the team is fine with it, I can make those changes but using those two tags will indicate using HTML tags for the complete details section.

**Comment by eggplantzzz** [comment]
You shouldn't have to use any other html besides those two tags. It should look something like this
```
## Testing the smart contract with Solidity
<details>
<summary>Expand this section</summary>
 //// normally formatted content for the tutorial
////
</details>

## Testing the smart contract with JavaScript
<details>
<summary>Expand this section</summary>
//// more content
////
</details>
```

Does that make sense? Or did I misunderstand what you meant? You shouldn't have to add any html explicitly in those sections besides the `<details>` and `<summary>` tags.

**Comment by AemieJ** [comment]
@eggplantzzz I implemented in the same manner however when I ran `yarn dev`, all content contained within the **details** tag didn't appear as stated with the markdown style.

**Comment by eggplantzzz** [comment]
Can you show me what you did? Do you have a side branch where you can demonstrate? I did this with no problems to get the screenshot you see above. Maybe I can figure out what's wrong.

**Comment by AemieJ** [comment]
@eggplantzzz it was an issue from my side, I have pushed the changes and it works perfectly.

**Comment by eggplantzzz** [comment]
Great! Thanks @AemieJ, the progress is looking good!

**Comment by AemieJ** [comment]
@eggplantzzz thanks, and let me know if any further changes are required to be implemented,

File: src/tutorials/pet-shop.md [hunk 11]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
    ```
 
 Note the **memory** attribute on `adopters`. The memory attribute tells Solidity to temporarily store the value in memory, rather than saving it to the contract's storage. Since `adopters` is an array, and we know from the first adoption test that we adopted pet `expectedPetId`, we compare the testing contracts address with location `expectedPetId` in the array.
+</details>
+
+## Testing the smart contract using Javascript 
+
+<details>
+<summary> Expand This Section </summary>
+Truffle is very flexible when it comes to smart contract testing, in that tests can be written either in JavaScript or Solidity. In this tutorial, we'll be writing our tests in Javascript using chai and mocha library.<br/>
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:385]: "Javascript" should be "JavaScript".

Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:385]: This should also say "...using the Chai and Mocha libraries.".

File: src/tutorials/pet-shop.md [hunk 12]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+2. Add the following content to the `testAdoption.test.js` file:<br/>
+
+  ```
+  const Adoption = require("../contracts/Adoption.sol");
+  
+  contract("Adoption", (accounts) => {
+    let adoption;
+    let expectedPetId;
+    
+    before(async() => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:397]: Can you put a space between "async" and "()"? So it would be `before(async () => {`. I think this is more consistent with the style used elsewhere in the docs.

File: src/tutorials/pet-shop.md [hunk 13]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+        adoption = await Adoption.deployed();
+    });
+
+    describe("adopting a pet and retrieving account addresses", async() => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:401]: Same comment about a space between "async" and "()" in the function.

File: src/tutorials/pet-shop.md [hunk 14]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+      before("adopt a pet using accounts[0]", async () => {
+      await adoption.adopt(8, { from: accounts[0] });
+      expectedAdopter = accounts[0];
+    });
+    });
+  });
+
+  ```
+  We start the contract by importing : 
+  * ```Adoption.sol```:The smart contract we want to test
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:411]: We should remove the note about importing chai. Later we could tell the reader that Truffle automatically imports it for you and that is why we can use `assert`. So this probably should not be a list any more if we remove the line about chai.

So I think it should be "We begin our test by importing our `Adoption` contract using `artifacts.require`."

File: src/tutorials/pet-shop.md [hunk 15]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  * ```chai```: Gives us various assertions to use in our tests to be used for equality, inequality or emptiness to return a pass/fail from our test.
+
+  **Note**: When writing the contract, we make use of the callback function ```accounts``` that provides us with the accounts in the development or the test network used.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:414]: This should read something like "When writing this test, our callback function takes an argument (`accounts`). This variable provides us with the accounts available on the network that we are using for testing."

File: src/tutorials/pet-shop.md [hunk 16]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+    let adoption;
+    let expectedPetId;
+    
+    before(async() => {
+        adoption = await Adoption.deployed();
+    });
+
+    describe("adopting a pet and retrieving account addresses", async() => {
+      before("adopt a pet using accounts[0]", async () => {
+      await adoption.adopt(8, { from: accounts[0] });
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:403]: This should be indented 2 spaces as well as the next two lines.

File: src/tutorials/pet-shop.md [hunk 17]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  ```
+  We start the contract by importing : 
+  * ```Adoption.sol```:The smart contract we want to test
+  * ```chai```: Gives us various assertions to use in our tests to be used for equality, inequality or emptiness to return a pass/fail from our test.
+
+  **Note**: When writing the contract, we make use of the callback function ```accounts``` that provides us with the accounts in the development or the test network used.
+
+  Then, we make use of async functions : 
+  * To deploy the contract as soon as the test begins for the contract
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:417]: This is not technically true because Truffle runs the migrations for you. I think we should get rid of this line. Maybe we should make a comment about the fact that Truffle does this for you somewhere in the test. Do you have an idea of where a good place might be?

I suppose this section should explain the `before` portion of the code. It could explain that everything in the `before` portion is test setup and that we are adopting the pet with id 8 using the address in `accounts[0]`. This allows us to later check and see that indeed pet with id 8 was adopted by `accounts[0]`.

File: src/tutorials/pet-shop.md [hunk 18]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  **Note**: When writing the contract, we make use of the callback function ```accounts``` that provides us with the accounts in the development or the test network used.
+
+  Then, we make use of async functions : 
+  * To deploy the contract as soon as the test begins for the contract 
+  * To use the asynchronous feature to get the address of the contract and expectedAdopter
+
+  ### Testing the adopt() function 
+
+  To test the ```adopt()``` function, recall that upon success it returns the given ```petId```. We can ensure an ID was returned and that it's correct by comparing the return value of ```adopt()``` to the ID we passed in.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:422]: I think all these instances of "adopt()" should be styled as mentioned above.

File: src/tutorials/pet-shop.md [hunk 19]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  * ```Adoption.sol```:The smart contract we want to test
+  * ```chai```: Gives us various assertions to use in our tests to be used for equality, inequality or emptiness to return a pass/fail from our test.
+
+  **Note**: When writing the contract, we make use of the callback function ```accounts``` that provides us with the accounts in the development or the test network used.
+
+  Then, we make use of async functions : 
+  * To deploy the contract as soon as the test begins for the contract 
+  * To use the asynchronous feature to get the address of the contract and expectedAdopter
+
+  ### Testing the adopt() function
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:420]: I think when talking about a function you should omit the "()" because that indicates a function call rather than the function itself. And you can probably use backticks (``) around your function name to style it. 

So like "Testing the `adopt` function"

Usually the backticks are used to indicate a variable or function name or something that is code.

File: src/tutorials/pet-shop.md [hunk 20]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  });
+
+  ```
+  We start the contract by importing : 
+  * ```Adoption.sol```:The smart contract we want to test
+  * ```chai```: Gives us various assertions to use in our tests to be used for equality, inequality or emptiness to return a pass/fail from our test.
+
+  **Note**: When writing the contract, we make use of the callback function ```accounts``` that provides us with the accounts in the development or the test network used.
+
+  Then, we make use of async functions :
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:416]: If you want to make a note that we need to use `await` because transaction methods are asynchronous then that might be helpful. What do you think?

File: src/tutorials/pet-shop.md [hunk 21]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  Then, we make use of async functions : 
+  * To deploy the contract as soon as the test begins for the contract 
+  * To use the asynchronous feature to get the address of the contract and expectedAdopter
+
+  ### Testing the adopt() function 
+
+  To test the ```adopt()``` function, recall that upon success it returns the given ```petId```. We can ensure an ID was returned and that it's correct by comparing the return value of ```adopt()``` to the ID we passed in.
+
+  1. Add the following function within the ```testAdoption.test.js``` test file, after the declaration of ```describe```.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:424]: I think this reads better as something like "After the `before` block of code, add the following function to `testAdoption.test.js`."

Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:424]: Also note, when you just want to have the backtick style inline, you only need to use one backtick. So like `testAdoption.test.js`. I think the 3 backticks will put it on its own line.

File: src/tutorials/pet-shop.md [hunk 22]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+    it("can fetch the address of an owner by pet id", async () => {
+      const adopter = await adoption.adopters(8);
+      assert.equal(adopter, expectedAdopter, "The owner of the adopted pet should be the first account.");
+    });
+  });
+  ```
+
+  Things to notice:
+
+  * We call the smart contract we declared earlier with the ID of expectedPetId.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:442]: This changed so it probably should be something like "We call smart contract method `adopters` to see what address adopted the pet with ID 8"

File: src/tutorials/pet-shop.md [hunk 23]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  * When called the function within the smart contract, we send metadata for the solidity contract to know that ```msg.sender``` is ```accounts[0]```.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:443]: I'm not sure if this is something relevant to mention here.

File: src/tutorials/pet-shop.md [hunk 24]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  * Finally, we pass the actual value, the expected value and a failure message (which gets printed to the console if the test does not pass) to assert.equal().
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:444]: We should probably just mention here what we are expecting. You could also mention that Truffle imports Chai for you so we have `assert` available to us in the test environment.

File: src/tutorials/pet-shop.md [hunk 25]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  ### Testing retrieval of a single pet's owner
+
+  1. Add this function below the previously added function in ```testAdoption.test.js```.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:448]: We probably should increment these numbers. They all read `1` now. It is probably better to have `1`, `2`, and `3`.

File: src/tutorials/pet-shop.md [hunk 26]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  });
+  ```
+
+  Things to notice:
+
+  * We call the smart contract we declared earlier with the ID of expectedPetId.
+  * When called the function within the smart contract, we send metadata for the solidity contract to know that ```msg.sender``` is ```accounts[0]```.
+  * Finally, we pass the actual value, the expected value and a failure message (which gets printed to the console if the test does not pass) to assert.equal().
+
+  ### Testing retrieval of a single pet's owner
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:446]: Maybe this section should just be a part of the previous section. Then it would have to ben named something like "Testing the `adopt` function and retrieving adopters' addresses". What do you think?

File: src/tutorials/pet-shop.md [hunk 27]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  1. Add this function below the previously added function in ```testAdoption.test.js```.
+
+  ```
+  it("can fetch the address of an owner by pet id", async () => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:451]: This should be the test testing the `getAdopters` function. The following test: 
```
    it("can fetch the collection of all pet owners addresses", async () => {
      const adopters = await adoption.getAdopters();
      assert.equal(adopters[8], expectedAdopter, "The owner of the adopted pet should be in the collection.");
    });
```

Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:451]: We already added this code in the tutorial. I think line 450-455 can be removed. You can probably add line 457 to the explanation above where we added this test.

File: src/tutorials/pet-shop.md [hunk 28]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+    const adopter = await adoption.adopters(8);
+    assert.equal(adopter, expectedAdopter, "The owner of the adopted pet should be the first account.");
+  });
+  ```
+
+  After getting the adopter address stored by the adoption contract, as assert equality as we did above.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:457]: "adopter's address"

File: src/tutorials/pet-shop.md [hunk 29]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+
+  ### Testing retrieval of all pet owners
+
+  Since arrays can only return a single value given a single key, we create our own getter for the entire array.
+
+  1. Add this function below the previously added function in ```testAdoption.test.js```.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:463]: This should probably be a different number based on whatever the previous number is.

File: src/tutorials/pet-shop.md [hunk 30]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+  
+  ```
+  it("can fetch the collection of all pet owners addresses", async () => {
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:466]: Sorry I made a grammar mistake. This should be "...of all pet owners' addresses"

File: src/tutorials/pet-shop.md [hunk 31]
```diff
@@ -374,6 +376,100 @@ Since arrays can only return a single value given a single key, we create our ow
...
+    const adopters = await adoption.getAdopters();
+    assert.equal(adopters[8], expectedAdopter, "The owner of the adopted pet should be in the collection.");
+  });
+  ```
+  Since adopters is an array, and we know from the first adoption test that we adopted pet expectedPetId, we compare the testing contracts address with location expectedPetId in the array.
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:471]: "the pet with ID 8, we are comparing the contract's address with the address that we expect to find."

**Comment by eggplantzzz** [comment]
I made another round of comments @AemieJ. If you end up getting tired, bored, or don't feel like continuing to work on this, feel free to let me know and I'll take over. Otherwise I'm happy to have the help!

**Comment by AemieJ** [comment]
@eggplantzzz I have no problem working on it, will look through the comments and make required changes in the upcoming days,

File: src/tutorials/pet-shop.md (hunk 25 above)
Comment (AemieJ) [reply src/tutorials/pet-shop.md]: It's not necessary but if we want to increment it, I will have to make similar changes for Test With Solidity section as well.

File: src/tutorials/pet-shop.md (hunk 26 above)
Comment (AemieJ) [reply src/tutorials/pet-shop.md]: Yes, it is the same section as the previous one so I have removed this section.

File: src/tutorials/pet-shop.md (hunk 29 above)
Comment (AemieJ) [reply src/tutorials/pet-shop.md]: It's not necessary but if we want to increment it, I will have to make similar changes for Test With Solidity section as well.

File: src/tutorials/pet-shop.md [hunk 32]
```diff
@@ -378,50 +378,50 @@ Since arrays can only return a single value given a single key, we create our ow
...
+        expectedAdopter = accounts[0];
+      });
     });
   });
 
   ```
   We start the contract by importing : 
-  * ```Adoption.sol```:The smart contract we want to test
-  * ```chai```: Gives us various assertions to use in our tests to be used for equality, inequality or emptiness to return a pass/fail from our test.
+  * `Adoption.sol`: The smart contract we want to test
```
Comment (eggplantzzz) [inline src/tutorials/pet-shop.md:411]: I suppose this should just be the contract name instead of the filename.

**Comment by eggplantzzz** [comment]
Ok @AemieJ, I think this is almost ready. Let me get one of my colleagues to go over it and we can get it merged. Can you make that one edit to the contract name on line 411? Thanks again for the work!
